src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/rollup/RollupMapperService.kt

```diff
@@ -15,16 +15,272 @@
+            currMap = nextMap as Map<*, *>
+        }
+
+        return true
+    }
+
+    private suspend fun jobExistsInRollupIndex(rollup: Rollup): RollupJobValidationResult {
+        val res = when (val getMappingsResult = getMappings(rollup.targetIndex)) {
+            is GetMappingsResult.Success -> getMappingsResult.response
+            is GetMappingsResult.Failure ->
+                return RollupJobValidationResult.Failure(getMappingsResult.message, getMappingsResult.cause)
+        }
+
+        val indexMapping: MappingMetadata = res.mappings[rollup.targetIndex][_DOC]
+
+        return if (((indexMapping.sourceAsMap?.get(_META) as Map<*, *>?)?.get(ROLLUPS) as Map<*, *>?)?.containsKey(rollup.id) == true) {
+            RollupJobValidationResult.Valid
+        } else {
+            RollupJobValidationResult.Invalid("Rollup job [${rollup.id}] does not exist in rollup index [${rollup.targetIndex}]")
+        }
+    }
+
+    @Suppress("ReturnCount")
+    private suspend fun getMappings(index: String): GetMappingsResult {
+        val errorMessage = "Failed to get mappings for index [$index]"
+        try {
+            val req = GetMappingsRequest().indices(index)
+            val res: GetMappingsResponse? = client.admin().indices().suspendUntil { getMappings(req, it) }
+            return if (res == null) {
+                GetMappingsResult.Failure(cause = IllegalStateException("GetMappingsResponse for index [$index] was null"))
+            } else {
+                GetMappingsResult.Success(res)
+            }
+        } catch (e: RemoteTransportException) {
+            val unwrappedException = ExceptionsHelper.unwrapCause(e) as Exception
+            logger.error(errorMessage, unwrappedException)
+            return GetMappingsResult.Failure(errorMessage, unwrappedException)
+        } catch (e: Exception) {
+            logger.error(errorMessage, e)
+            return GetMappingsResult.Failure(errorMessage, e)
+        }
+    }
+
+    fun isRollupIndex(index: String): Boolean = RollupSettings.ROLLUP_INDEX.get(clusterService.state().metadata.index(index).settings)
+
+    fun indexExists(index: String): Boolean = clusterService.state().routingTable.hasIndex(index)
+
+    // TODO: error handling - can RemoteTransportException happen here?
+    // TODO: The use of the master transport action UpdateRollupMappingAction will prevent
+    //   overwriting an existing rollup job _meta by checking for the job id
+    //   but there is still a race condition if two jobs are added at the same time for the
+    //   same target index. There is a small time window after get mapping and put mappings
+    //   where they can both get the same mapping state and only add their own job, meaning one
+    //   of the jobs won't be added to the target index _meta
+    @Suppress("BlockingMethodInNonBlockingContext", "ReturnCount")
+    private suspend fun updateRollupIndexMappings(rollup: Rollup): RollupJobValidationResult {
```

Comment:
Optional - TODO: disable and not update job in target index if it clashes with other existing jobs, e.g. if the job1 is using fieldA as metric then job2 cannot use fieldA as dimension.